nlmod/mdims/mlayers.py

```diff
@@ -815,198 +775,158 @@ def fill_top_bot_kh_kv_at_mask(model_ds, fill_mask, gridtype="structured"):
+    # TODO: set idomain above and below the first active layer to 0
 
-        # update idomain on adjusted tops and bots
-        model_ds["thickness"], _ = calculate_thickness(model_ds)
-        model_ds["idomain"] = update_idomain_from_thickness(
-            model_ds["idomain"], model_ds["thickness"], model_ds["northsea"]
-        )
-        model_ds["first_active_layer"] = mgrid.get_first_active_layer_from_idomain(
-            model_ds["idomain"]
-        )
-
-    else:
-        model_ds["thickness"], _ = calculate_thickness(model_ds)
-        model_ds["first_active_layer"] = mgrid.get_first_active_layer_from_idomain(
-            model_ds["idomain"]
-        )
+    ds["first_active_layer"] = get_first_active_layer_from_idomain(ds["idomain"],
+                                                                   nodata=nodata)
 
-    return model_ds
+    ds.attrs["nodata"] = nodata
+    return ds
 
 
-def add_idomain_from_bottom_to_dataset(bottom, model_ds, nodata=-999):
-    """add idomain and first_active_layer to model_ds The active layers are
-    defined as the layers where the bottom is not nan.
+def get_first_active_layer_from_idomain(idomain, nodata=-999):
+    """get the first active layer in each cell from the idomain.
 
     Parameters
     ----------
-    bottom : xarray.DataArray
-        DataArray with bottom values of each layer. Nan values indicate
-        inactive cells.
-    model_ds : xarray.Dataset
-        dataset with model data where idomain and first_active_layer
-        are added to.
+    idomain : xr.DataArray
+        idomain. Shape can be (layer, y, x) or (layer, icell2d)
     nodata : int, optional
-        nodata value used in integer arrays. For float arrays np.nan is use as
-        nodata value. The default is -999.
+        nodata value. used for cells that are inactive in all layers.
+        The default is -999.
 
     Returns
     -------
-    model_ds : xarray.Dataset
-        dataset with model data including idomain and first_active_layer
+    first_active_layer : xr.DataArray
+        raster in which each cell has the zero based number of the first
+        active layer. Shape can be (y, x) or (icell2d)
     """
-    logger.info("get active cells (idomain) from bottom DataArray")
-
-    idomain = xr.where(bottom.isnull(), -1, 1)
-
-    # if the top cell is inactive set idomain = 0, for other inactive cells
-    # set idomain = -1
-    idomain[0] = xr.where(idomain[0] == -1, 0, idomain[0])
-    for i in range(1, bottom.shape[0]):
-        idomain[i] = xr.where((idomain[i - 1] == 0) & (idomain[i] == -1), 0, idomain[i])
-
-    model_ds["idomain"] = idomain
-    model_ds["first_active_layer"] = mgrid.get_first_active_layer_from_idomain(
-        idomain, nodata=nodata
-    )
+    logger.info("get first active modellayer for each cell in idomain")
 
-    model_ds.attrs["nodata"] = nodata
+    first_active_layer = xr.where(idomain[0] == 1, 0, nodata)
```

Comment:
is this todo relevant, or already done?

Reply:
What todo?

Reply:
Oh, it is for the line:
_# TODO: set idomain above and below the first active layer to 0_
Right now idomain of non-existent cells is set to -1 (pass-through), even it it is the top or the lowest layer. I think this does not cause any problems with MODFLOW, but it may be better to set idomain in these cells to 0. This is what the todo is about.

Reply:
Ok no problem, just checking if it was not done already.